src/item.h

```diff
@@ -327,95 +318,106 @@ class ItemAttributes
 			};
 		};
 
+		typedef boost::variant<boost::blank, std::string, int64_t, double, bool> VariantAttribute;
+
 		struct CustomAttribute
 		{
-			union {
-				std::string* string;
-				int64_t integer;
-				double doublen;
-				bool boolean;
-			} value;
+			VariantAttribute value;
 			customAttrTypes type = ATTR_NO_TYPE;
 
 			explicit CustomAttribute(customAttrTypes type) : type(type) {
-				memset(&value, 0, sizeof(value));
+				switch (type) {
+					case ATTR_STRING_TYPE: {
+						value = std::string();
+						break;
+					}
+
+					case ATTR_INTEGER_TYPE: {
+						value = (int64_t)0;
+						break;
+					}
+
+					case ATTR_DOUBLE_TYPE: {
+						value = (double)0;
+						break;
+					}
+
+					case ATTR_BOOLEAN_TYPE: {
+						value = false;
+						break;
+					}
+
+					default: {
+						break;
+					}
+				}
 			}
 
 			CustomAttribute(const std::string& v) {
 				type = ATTR_STRING_TYPE;
-				value.string = new std::string(v);
+				value = std::string(v);
 			}
 
 			CustomAttribute(const char* v) {
 				type = ATTR_STRING_TYPE;
-				value.string = new std::string(v);
+				value = std::string(v);
 			}
 
 			CustomAttribute(int64_t v) {
 				type = ATTR_INTEGER_TYPE;
-				value.integer = v;
+				value = v;
 			}
 
 			CustomAttribute(double v) {
 				type = ATTR_DOUBLE_TYPE;
-				value.doublen = v;
+				value = v;
 			}
 
 			CustomAttribute(bool v) {
 				type = ATTR_BOOLEAN_TYPE;
-				value.boolean = v;
+				value = v;
 			}
 
 			CustomAttribute(const CustomAttribute& i) {
```

Comment:
I don't think you need to define the copy and move constructors/assignment operators. The defaults are fine.